engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java

```diff
@@ -443,15 +447,18 @@ private String interpolate(String messageTemplate,
 		}
 	}
 
-	private boolean isAlreadyValidatedForPath(Object value, PathImpl path) {
-		Set<PathImpl> pathSet = processedPathsPerBean.get( value );
+	private boolean isAlreadyValidatedForPathAndTypeParameter(Object value, PathImpl path, TypeVariable<?> typeParameter) {
+		Set<PathAndTypeParameter> pathSet = processedPathsPerBean.get( value );
 		if ( pathSet == null ) {
 			return false;
 		}
 
-		for ( PathImpl p : pathSet ) {
-			if ( path.isRootPath() || p.isRootPath() || isSubPathOf( path, p ) || isSubPathOf( p, path ) ) {
-				return true;
+		for ( PathAndTypeParameter p : pathSet ) {
+			if ( path.isRootPath() || p.path.isRootPath() || isSubPathOf( path, p.path ) || isSubPathOf( p.path, path ) ) {
+				if ( typeParameter == p.typeParameter ||
+						( p.typeParameter == null && ( typeParameter == AnnotatedObject.INSTANCE || typeParameter == ArrayElement.INSTANCE ) ) ) {
```

Comment:
It has become obsolete with my latest commit (going to push it asap). It was needed to prevent repeated validation of an object which was part of the graph (e.g. the validated bean) _and_ contained within an array visited during cascaded validation.